mmaction/models/heads/ssn_head.py

```diff
@@ -0,0 +1,414 @@
+        else:
+            self.dropout = None
+
+        consensus_ = consensus.copy()
+        consensus_type = consensus_.pop('type')
+        if consensus_type == 'STPPTrain':
+            self.consensus = STPPTrain(**consensus_)
+        elif consensus_type == 'STPPTest':
+            consensus_['num_classes'] = self.num_classes
+            self.consensus = STPPTest(**consensus_)
+
+        self.in_channels_complete = (
+            self.consensus.feat_multiplier * in_channels_activity)
+        self.activity_fc = nn.Linear(in_channels_activity, num_classes + 1)
+        self.completeness_fc = nn.Linear(self.in_channels_complete,
+                                         num_classes)
+        if self.with_regression:
+            self.regressor_fc = nn.Linear(self.in_channels_complete,
+                                          num_classes * 2)
+
+    def init_weights(self):
+        """Initiate the parameters from scratch."""
+        normal_init(self.activity_fc, std=self.init_std)
+        normal_init(self.completeness_fc, std=self.init_std)
+        if self.with_regression:
+            normal_init(self.regressor_fc, std=self.init_std)
+
+    def prepare_test_fc(self, stpp_feat_multiplier):
+        """Reorganize the shape of fully connected layer at testing, in order
+        to improve testing efficiency.
+
+        Args:
+            stpp_feat_multiplier (int): Total number of parts.
+
+        Returns:
+            bool: Whether the shape transformation is ready for testing.
+        """
+
+        in_features = self.activity_fc.in_features
+        out_features = (
+            self.activity_fc.out_features +
+            self.completeness_fc.out_features * stpp_feat_multiplier)
+        if self.with_regression:
+            out_features += (
+                self.regressor_fc.out_features * stpp_feat_multiplier)
+        self.test_fc = nn.Linear(in_features, out_features)
+
+        # Fetch weight and bias of the reorganized fc.
+        complete_weight = self.completeness_fc.weight.data.view(
+            self.completeness_fc.out_features, stpp_feat_multiplier,
+            self.activity_fc.in_features).transpose(0, 1).contiguous().view(
+                -1, self.activity_fc.in_features)
+        complete_bias = self.completeness_fc.bias.data.view(1, -1).expand(
+            stpp_feat_multiplier, self.completeness_fc.out_features
+        ).contiguous().view(-1) / stpp_feat_multiplier
+
+        weight = torch.cat((self.activity_fc.weight.data, complete_weight))
+        bias = torch.cat((self.activity_fc.bias.data, complete_bias))
+
+        if self.with_regression:
+            reg_weight = self.regressor_fc.weight.data.view(
+                self.regressor_fc.out_features,
+                stpp_feat_multiplier, self.activity_fc.in_features).transpose(
+                    0, 1).contiguous().view(-1, self.activity_fc.in_features)
+            reg_bias = self.regressor_fc.bias.data.view(1, -1).expand(
```

Comment:
is expand().contiguous() equivalent to repeat() ?

Reply:
I tried it, but failed, there was an shape error.